usr/src/uts/common/os/sysent.c

```diff
@@ -434,329 +429,277 @@ struct sysent nosys_ent = SYSENT_NOSYS();
  */
 struct sysent sysent[NSYSCALL] =
 {
-	/*  0 */ IF_LP64(
-			SYSENT_NOSYS(),
-			SYSENT_C("indir",	indir,		1)),
-	/*  1 */ SYSENT_CI("exit",		rexit,		1),
-	/*  2 */ SYSENT_CI("psecflags",		psecflags,	3),
-	/*  3 */ SYSENT_CL("read",		read,		3),
-	/*  4 */ SYSENT_CL("write",		write,		3),
-	/*  5 */ SYSENT_CI("open",		open,		3),
-	/*  6 */ SYSENT_CI("close",		close,		1),
-	/*  7 */ SYSENT_CI("linkat",		linkat,		5),
-	/*  8 */ SYSENT_LOADABLE(),			/* (was creat) */
-	/*  9 */ SYSENT_CI("link",		link,		2),
-	/* 10 */ SYSENT_CI("unlink",		unlink,		1),
-	/* 11 */ SYSENT_CI("symlinkat",		symlinkat,	3),
-	/* 12 */ SYSENT_CI("chdir",		chdir,		1),
-	/* 13 */ SYSENT_CL("time",		gtime,		0),
-	/* 14 */ SYSENT_CI("mknod",		mknod,		3),
-	/* 15 */ SYSENT_CI("chmod",		chmod,		2),
-	/* 16 */ SYSENT_CI("chown",		chown,		3),
-	/* 17 */ SYSENT_CI("brk",		brk,		1),
-	/* 18 */ SYSENT_CI("stat",		stat,		2),
-	/* 19 */ IF_LP64(
-			SYSENT_CL("lseek",	lseek64,	3),
-			SYSENT_CL("lseek",	lseek32,	3)),
-	/* 20 */ SYSENT_2CI("getpid",		getpid,		0),
-	/* 21 */ SYSENT_AP("mount",		mount,		8),
-	/* 22 */ SYSENT_CL("readlinkat",	readlinkat,	4),
-	/* 23 */ SYSENT_CI("setuid",		setuid,		1),
-	/* 24 */ SYSENT_2CI("getuid",		getuid,		0),
-	/* 25 */ SYSENT_CI("stime",		stime,		1),
-	/* 26 */ SYSENT_CL("pcsample",		pcsample,	2),
-	/* 27 */ SYSENT_CI("alarm",		alarm,		1),
-	/* 28 */ SYSENT_CI("fstat",		fstat,		2),
-	/* 29 */ SYSENT_CI("pause",		pause,		0),
-	/* 30 */ SYSENT_LOADABLE(),			/* (was utime) */
-	/* 31 */ SYSENT_CI("stty",		stty,		2),
-	/* 32 */ SYSENT_CI("gtty",		gtty,		2),
-	/* 33 */ SYSENT_CI("access",		access,		2),
-	/* 34 */ SYSENT_CI("nice",		nice,		1),
-	/* 35 */ IF_LP64(
-			SYSENT_NOSYS(),
-			SYSENT_CI("statfs",	statfs32,	4)),
-	/* 36 */ SYSENT_CI("sync",		syssync,	0),
-	/* 37 */ SYSENT_CI("kill",		kill,		2),
-	/* 38 */ IF_LP64(
-			SYSENT_NOSYS(),
-			SYSENT_CI("fstatfs",	fstatfs32,	4)),
-	/* 39 */ SYSENT_CI("setpgrp",		setpgrp,	3),
-	/* 40 */ SYSENT_CI("uucopystr",		uucopystr,	3),
-	/* 41 */ SYSENT_LOADABLE(),			/* (was dup) */
-	/* 42 */ SYSENT_LOADABLE(),			/* pipe */
-	/* 43 */ SYSENT_CL("times",		times,		1),
-	/* 44 */ SYSENT_CI("prof",		profil,		4),
-	/* 45 */ SYSENT_CI("faccessat",		faccessat,	4),
-	/* 46 */ SYSENT_CI("setgid",		setgid,		1),
-	/* 47 */ SYSENT_2CI("getgid",		getgid,		0),
-	/* 48 */ SYSENT_CI("mknodat",		mknodat,	4),
-	/* 49 */ SYSENT_LOADABLE(),			/* msgsys */
-	/* 50 */ IF_x86(
+	[0]	= SYSENT_NOSYS(),	/* Was indirect syscall on 32bit */
```

Comment:
Isn't this an indirect system call now on ARM?

Reply:
See the prior discussion in https://github.com/richlowe/illumos-gate/pull/43#discussion_r1327793879

syscall 0 is indirect, but implemented as such in libc on the 64bit kernels, and is just `nosys` in the kernel table.  Removing the lingering 32/64 ifdefs leaves it a constant nosys.  The `indir` function remains, but itself also returns `nosys` on every platform except 32bit SPARC.

Reply:
Can I trouble you to point me to the aarch64 implementation. I'm having trouble understanding how if we come in on vector zero as indirect that it gets translated to a useful function if the sysent is `nosys`. I guess that's happening somehow before we get to the sysent table similar to what we do for x86 in the assembly files?

Reply:
If we get a trap from a lower level, we check whether it is `svc` (at the [`from_lower_el_aarch64_sync`](https://github.com/richlowe/illumos-gate/blob/arm64-gate/usr/src/uts/armv8/ml/exceptions.S#L103) label), and branch to [`svc_handler` ](https://github.com/richlowe/illumos-gate/blob/arm64-gate/usr/src/uts/armv8/ml/exceptions.S#L174).

`svc_handler`  masks out the syscall# and if it is 0, replaces it with the value from `%x9` (actually `%w9` the 32bit view of it).  This is [here](https://github.com/richlowe/illumos-gate/blob/arm64-gate/usr/src/uts/armv8/ml/exceptions.S#L180C1-L182C13).

This is basically the same as in `sys_sysenter` etc. on x86, except they don't have the syscall# encoded in the instruction (ever) and always indirect through a register (`%rax`/`%eax`).

Reply:
OK, thanks. I think that was the last piece I was confused with on this.